File: src/kangaroo_helper_ffi.mjs
export function metadata(_value) {}

export function serial() {}

export function skip(reason) {
  const error = new Error(`kangaroo skip: ${reason}`);
  error.kangaroo_skip = true;
  error.reason = reason;
  throw error;
}

const isPromise = (value) => value && typeof value.then === "function";

export function fixture(setup, teardown, body) {
  const resource = setup();
  if (isPromise(resource)) {
    return resource.then((resolved) => fixtureValue(resolved, teardown, body));
  }
  return fixtureValue(resource, teardown, body);
}

function fixtureValue(resource, teardown, body) {
  let bodyValue;
  try {
    bodyValue = body(resource);
  } catch (bodyError) {
    return finishFailure(resource, teardown, bodyError);
  }

  if (isPromise(bodyValue)) {
    return bodyValue.then(
      async (value) => {
        await teardown(resource);
        return value;
      },
      (bodyError) => finishFailureAsync(resource, teardown, bodyError),
    );
  }

  const cleanup = teardown(resource);
  if (isPromise(cleanup)) return cleanup.then(() => bodyValue);
  return bodyValue;
}

function finishFailure(resource, teardown, bodyError) {
  try {
    const cleanup = teardown(resource);
    if (isPromise(cleanup)) {
      return cleanup.then(
        () => Promise.reject(bodyError),
        (cleanupError) => Promise.reject(combined(bodyError, cleanupError)),
      );
    }
  } catch (cleanupError) {
    throw combined(bodyError, cleanupError);
  }
  throw bodyError;
}

async function finishFailureAsync(resource, teardown, bodyError) {
  try {
    await teardown(resource);
  } catch (cleanupError) {
    throw combined(bodyError, cleanupError);
  }
  throw bodyError;
}

function combined(bodyError, cleanupError) {
  return new AggregateError(
    [bodyError, cleanupError],
    `fixture body failed: ${message(bodyError)}; teardown failed: ${message(cleanupError)}`,
  );
}

function message(error) {
  return error && error.message ? error.message : String(error);
}

File: src/kangaroo_helper_ffi.erl
-module(kangaroo_helper_ffi).
-export([metadata/1, serial/0, skip/1, fixture/3]).

metadata(_Value) -> nil.

serial() -> nil.

skip(Reason) ->
    erlang:error(#{kangaroo_error => skip, reason => Reason}).

fixture(Setup, Teardown, Body) ->
    case capture(Setup) of
        {error, Class, Reason, Stack} ->
            erlang:raise(Class, Reason, Stack);
        {ok, Resource} ->
            BodyResult = capture(fun() -> Body(Resource) end),
            TeardownResult = capture(fun() -> Teardown(Resource) end),
            combine(BodyResult, TeardownResult)
    end.

capture(Fun) ->
    try {ok, Fun()}
    catch Class:Reason:Stack -> {error, Class, Reason, Stack}
    end.

combine({ok, Value}, {ok, _}) ->
    Value;
combine({error, Class, Reason, Stack}, {ok, _}) ->
    erlang:raise(Class, Reason, Stack);
combine({ok, _}, {error, Class, Reason, Stack}) ->
    erlang:raise(Class, Reason, Stack);
combine({error, _BodyClass, BodyReason, BodyStack},
        {error, _CleanupClass, CleanupReason, CleanupStack}) ->
    Message = iolist_to_binary([
        <<"fixture body failed: ">>, reason_text(BodyReason),
        <<"; teardown failed: ">>, reason_text(CleanupReason)
    ]),
    erlang:error(#{gleam_error => panic,
                   message => Message,
                   body_reason => BodyReason,
                   body_stack => BodyStack,
                   teardown_reason => CleanupReason,
                   teardown_stack => CleanupStack}).

reason_text(#{message := Message}) -> to_binary(Message);
reason_text(Reason) -> to_binary(io_lib:format("~0p", [Reason])).

to_binary(Value) when is_binary(Value) -> Value;
to_binary(Value) when is_list(Value) -> unicode:characters_to_binary(Value);
to_binary(Value) -> unicode:characters_to_binary(io_lib:format("~0p", [Value])).
